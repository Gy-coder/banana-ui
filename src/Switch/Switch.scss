$w: 44px;
$blue: #40a9ff;
.g-switch {
  height: $w / 2;
  width: $w;
  background: #c0ccda;
  border-radius: 999px;
  cursor: pointer;
  &.g-switch-checked {
    background: #20a0ff;
    > .g-switch-toggle-wrapper {
      margin-left: 50%;
    }
  }
  > .g-switch-toggle-wrapper {
    box-sizing: border-box;
    height: 100%;
    width: $w / 2;
    padding: $w / 25;
    transition: margin-left 250ms;
    margin-left: 0;
    > .g-switch-toggle {
      width: 100%;
      height: 100%;
      border-radius: 50%;
      background: white;
      display: inline-flex;
      justify-content: center;
      align-items: center;
      > .g-loadingIndicator {
        box-sizing: border-box;
        width: 14px;
        height: 14px;
        display: inline-block;
        border-radius: 8px;
        border-color: $blue $blue $blue transparent;
        border-style: solid;
        border-width: 2px;
        animation: g-spin 1s infinite linear;
      }
    }
  }
  &.disabled {
    background: grey;
    cursor: not-allowed;
    .g-loadingIndicator {
      box-sizing: border-box;
      width: 14px;
      height: 14px;
      display: inline-block;
      border-radius: 8px;
      border-color: grey grey grey transparent !important;
      border-style: solid;
      border-width: 2px;
      animation: g-spin 1s infinite linear;
    }
  }
}
