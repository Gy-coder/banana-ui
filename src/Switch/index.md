---
title: Swtich
nav:
  title: 数据录入
  path: /components
---

# Swtich

开关选择器。

# 何时使用

- 需要表示开关状态/两种状态之间的切换时；

## 代码范例

<code src="./demos/basic.tsx" />
<code src="./demos/click.tsx" />
<code src="./demos/disabled.tsx" />
<code src="./demos/loading.tsx" />

<API src="./Switch.tsx">
