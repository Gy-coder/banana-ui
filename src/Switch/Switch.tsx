import React, { useEffect, useState } from 'react';
import classnames from 'classnames';
import './Switch.scss';

interface Props {
  /**
   * 可以这样写属性描述
   * @description     Switch的类名
   */
  className?: string;
  /**
   * 可以这样写属性描述
   * @description     Switch的CSS样式
   */
  style?: React.CSSProperties;
  /**
   * 可以这样写属性描述
   * @description     Switch的选中值
   */
  value: boolean;
  /**
   * 可以这样写属性描述
   * @description     checked改变时的回调函数
   */
  onChange: (value: boolean) => void;
  /**
   * 可以这样写属性描述
   * @description     是否禁用
   * @default         false
   */
  disabled?: boolean;
  /**
   * 可以这样写属性描述
   * @description     是否处于加载中
   * @default         false
   */
  loading?: boolean;
}

const Switch: React.FC<Props> = (props) => {
  const { className, style, disabled, value, onChange, loading } = props;
  const classNames = classnames('g-switch', className, {
    'g-switch-checked': value,
    disabled: disabled,
  });
  const handleClick = () => {
    if (disabled) return;
    onChange(!value);
  };
  return (
    <div className={classNames} style={style} onClick={handleClick}>
      <div className="g-switch-toggle-wrapper">
        <div className="g-switch-toggle">
          {loading && <span className="g-loadingIndicator" />}
        </div>
      </div>
    </div>
  );
};

Switch.defaultProps = {
  loading: false,
  disabled: false,
};

export default Switch;
