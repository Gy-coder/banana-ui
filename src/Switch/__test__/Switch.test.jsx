import {
  act,
  render,
  cleanup,
  screen,
  fireEvent,
} from '@testing-library/react';
import '@testing-library/jest-dom';
import React, { useState } from 'react';
import Switch from '../Switch';

describe('测试Switch', () => {
  it('Switch可以正常渲染', () => {
    const { asFragment } = render(<Switch onChange={() => {}} value={false} />);
    expect(document.querySelector('.g-switch')).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });
  it('可以点击修改值', () => {
    const Temp = () => {
      const [value, setValue] = useState(false);
      return (
        <div data-testid="test">
          <Switch
            onChange={() => {
              setValue(!value);
            }}
            value={value}
          />
        </div>
      );
    };
    render(<Temp />);
    const el = screen.getByTestId('test').childNodes[0];
    expect(el.className).not.toContain('g-switch-checked');
    fireEvent.click(el);
    expect(el.className).toContain('g-switch-checked');
    fireEvent.click(el);
    expect(el.className).not.toContain('g-switch-checked');
  });
  it('可以disabled', () => {
    const Temp = () => {
      const [value, setValue] = useState(false);
      return (
        <div data-testid="test">
          <Switch
            disabled={true}
            onChange={() => {
              setValue(!value);
            }}
            value={value}
          />
        </div>
      );
    };
    render(<Temp />);
    const el = screen.getByTestId('test').childNodes[0];
    expect(el.className).toContain('disabled');
    expect(el.className).not.toContain('g-switch-checked');
    fireEvent.click(el);
    expect(el.className).not.toContain('g-switch-checked');
  });
  it('可以有loading', () => {
    render(<Switch onChange={() => {}} value={false} loading={true} />);
    expect(document.querySelector('.g-loadingIndicator')).toBeVisible();
  });
});
