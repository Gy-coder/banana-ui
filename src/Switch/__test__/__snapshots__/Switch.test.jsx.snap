// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试Switch Switch可以正常渲染 1`] = `
<DocumentFragment>
  <div
    class="g-switch"
  >
    <div
      class="g-switch-toggle-wrapper"
    >
      <div
        class="g-switch-toggle"
      />
    </div>
  </div>
</DocumentFragment>
`;
